### DISCUSSION TITLE
proposal: "throw?"

### DISCUSSION THREAD

**Discussion by TobiasKnauss** [dotnet/csharplang#3720] [Language Ideas]
The following code:
```
      var exception = CheckInputValue_ImageSize (i_sourceImageSize);
      if (exception != null)
        throw exception;
```
could be simplified as
```
   throw? CheckInputValue_ImageSize (i_sourceImageSize);
```
if there was an "throw?" command with the meaning "throw if not null".

**Comment by JasonBock** [comment]
Why would I return an exception from a method invocation? This seems like such a small corner case that it doesn't seem worth it to put time into adding this.

**Comment by svick** [comment]
Why do you use this pattern instead of directly throwing from `CheckInputValue_ImageSize`?

**Comment by foxesknow** [comment]
How common a use-case is it that you're been returned an exception which, if not null, you want to throw? I suspect it's almost zero.

**Comment by AlgorithmsAreCool** [comment]
> Why would I return an exception from a method invocation? 

> Why do you use this pattern instead of directly throwing from CheckInputValue_ImageSize?

While I agree this is too niche to add to the language. It is a moderately common pattern to move exception creation into helper methods to either:
- Centralize exception logic
- Make the parent method more inline friendly

**Comment by PathogenDavid** [comment]
This is actually a pattern I've seen quite a bit. (It's an OK pattern for implementing a pair of TryXyz and Xyz methods when you don't care about the extra allocation of the exception in the TryXyz case.)

That being said, I agree with @AlgorithmsAreCool, this is pretty niche.

If you're finding yourself implementing this pattern a lot, [I think you're better off making an internal extension method for `Exception`](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AEQEsBDAczwgGcM7MP2YBhCABMYAQTycANgE9+7fgG5q1dngwwo8hYzJJGJegFEEYGAAchEPFb15Vj/tQDe1er/oB6f3oAbVYIGH4AOQgMACUYDBwDAF16ODh6Dn5OYAUYSXoMCHoAW04Aaxh6AGUAC04oWwAZHPp621slHz9AkOqMTjBygBUoQZgACXZJaTxU9LaMDFt+EEDudgxanGBmSBL/SRi8BP9cXXYSmH81fhwI/xIATiQADkoqP3odPQNFY1MRBQ9GGtSgEAA7gAKLZqSzWOwOPD0GAI+zsRwASm6X18AF4AHyFMGQlFopGaKgAXy0NAAzACzCCIhgvDj6LYoOwAG6cPSMqw2dGOTIQUE6bgASRKtjyV10fIxeChP3oCDQ9AgOAw3109FgdwUGGxnz83lNuO+ADN6FCEPQ8Xj6LQTZbfOa3biDTgjQ76NIrZwfRhKZ6vkQAOz0E4Q+gyKDcHDyjAAeW1KatsU4eG4MEFiKVUPkVwgVrtmI1ACJ7SUcIJozF6MAqrQAISVzGht00i249lfb2+p32gBUaq7lsj0Z9Cgn9B7X3ZRAZTFMqrC4pzKr1CFdZv7fg3tQl0tlMGTiscdo1Wp1qsHxoA/MxQeDoZ2D74pw+5wu/EuVxMJsIAgIxRiUI8JW3HV1U1bVdR1B890tQlRU3KUZTlGAFSRa84MQiJg0xb5+GnBRZ2oHtqFIOk2V7UgSHZD03XvQijTnRde3DMgnihAASStwMgrcSGIp1PGGFkXygCCxWPESb3gpCqQ1B8/U8B8qQ7Di/CYXiBOE7goVE9TJMENg5Kg0StI/Lj/zsr8eP4wSZMMqEXVMqShMsrdaEUgjDWNFT9TYnVxM07TP3oCRXFAmBmAAdS5PQoUrNyPKdSKHOMfTKwkqT0sxGzf0ooA=) IE:

```csharp
internal static class ExceptionExtensions
{
    [DoesNotReturn]
    // [StackTraceHidden] -- https://github.com/dotnet/runtime/issues/29681
    internal static void Throw(this Exception exception)
        => throw exception;
}
```

Then use it as:

```csharp
CheckInputValue_ImageSize(i_sourceImageSize)?.Throw();
```

**Comment by TobiasKnauss** [comment]
@PathogenDavid : That's exactly what I need... if the `[StackTraceHidden]` attribute existed.
@AlgorithmsAreCool : Yes, I try to centralize the exception creation to avoid boilerplate code.
@svick This method calls other methods for more checks, and I prefer the exception to be thrown not in a 3rd-level check method, but rather in the method that should do the actual work, so you have a better understanding where it comes from.

Another reason for implementing this pattern: It's easier to write unit tests for these check methods. Instead of a try-catch in the unit test, you can just check the returned exception object.

**Comment by theunrepentantgeek** [comment]
I'll frequently create factory methods to create exceptions, but the logic controlling whether they throw will always be in the main method. Using factory methods, and moving exception instantiation out of the main flow of the method, gives room for better (more informative) exceptions.

So, I'd write the original code thusly:

``` csharp
if (!ImageSizeIsValid(i_sourceImageSize) 
{
    throw InvalidImageSizeError(i_sourceImageSize);
}
```
where the factory method might look something like this:

``` csharp
private InvalidOperationException InvalidImageSizeError(int imageSize) 
{
    var result = new InvalidOperationException($"Image size {imageSize} is invalid");
    result.Data[nameof(imageSize)] = imageSize;
    return result;
}
```

With respect to the original proposal, I don't believe a `throw?` keyword (or anything semantically similar, regardless of syntax) would be a good idea as it encourages poor design.

**Comment by TobiasKnauss** [comment]
> I'll frequently create factory methods to create exceptions, but the logic controlling whether they throw will always be in the main method. Using factory methods, and moving exception instantiation out of the main flow of the method, gives room for better (more informative) exceptions.

I do use this approach at other places, but the problem is, that you'll have to double the logic if you want to provide additional information about what is wrong.
E.g. if the image size is checked against lower and upper limits, then the information, which limit was exceeded, must somehow be forwarded to the method that creates the exception. Thus it's easier for me to create the exception at the place where the check is done, and then just throw it if it exists.

**Comment by PathogenDavid** [comment]
@TobiasKnauss 
> That's exactly what I need... if the `[StackTraceHidden]` attribute existed.

The API was approved yesterday, so I like to hope it'll become available sooner rather than later.

**Comment by jnm2** [comment]
> The API was approved yesterday, so I like to hope it'll become available sooner rather than later.

Can't it work to declare an internal attribute in the correct namespace in source, just like with all compiler-recognized attributes?

**Comment by PathogenDavid** [comment]
> Can't it work to declare an internal attribute in the correct namespace in source, just like with all compiler-recognized attributes?

I was hoping so too. It's not a compiler thing, but the Visual Studio debugger does not appear to match it by shape either:

![image](https://user-images.githubusercontent.com/278957/88463278-398a0c80-ce77-11ea-9152-ce5b0ba98e1e.png)

Additionally, [`StackTrace` uses it directly](https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Private.CoreLib/src/System/Diagnostics/StackTrace.cs#L356-L369), so it definitely needs the actual type.

**Comment by jnm2** [comment]
I wonder if they would accept a PR to match by full name rather than by metadata token.
